psyneulink/core/components/functions/userdefinedfunction.py

```diff
@@ -29,6 +30,295 @@
+            ("scalar", "scalar"): lambda x,y: self.builder.fmul(x, y),
+            ("mat", "scalar"): lambda x,y: gen_inject_mat_scalar_mult(self.ctx, self.builder, x, y),
+            ("vec", "scalar"): lambda x,y: gen_inject_vec_scalar_mult(self.ctx, self.builder, x, y),
+            ("vec", "vec"): lambda x,y: gen_inject_vec_hadamard(self.ctx, self.builder, x, y),
+            ("mat", "mat"): lambda x,y: gen_inject_mat_hadamard(self.ctx, self.builder, x, y),
+        }
+        return lambda x,y: self._generate_binop(x, y, handlers)
+
+    def _parse_unop_arg(self, x):
+        x_type = self._get_type(x)
+        # unpack scalars from pointers
+        if x_type == 'scalar' and type(getattr(x, "type", x)) is pnlvm.ir.PointerType:
+            x = self.builder.load(x)
+
+        return x, x_type
+
+    def _generate_unop(self, x, handlers:dict):
+        x, x_type = self._parse_unop_arg(x)
+        return handlers[(x_type)](x)
+
+    def visit_USub(self, node):
+        from psyneulink.library.compositions.pytorchllvmhelper import gen_inject_mat_scalar_mult, gen_inject_vec_scalar_mult
+        handlers = {
+            ("scalar"): lambda x: pnlvm.helpers.fneg(self.builder, x),
+            # FIXME: Add generalized vec/matrix fneg
+            ("vec"): lambda x: gen_inject_vec_scalar_mult(self.ctx, self.builder, x, self.ctx.float_ty(-1)),
+            ("mat"): lambda x: gen_inject_mat_scalar_mult(self.ctx, self.builder, x, self.ctx.float_ty(-1)),
+        }
+        return lambda x: self._generate_unop(x, handlers)
+
+    def visit_Name(self, node):
+        return self.register[node.id]
+
+    def visit_Attribute(self, node):
+        val = self.visit(node.value)
+        return val[node.attr]
+
+    def visit_Num(self, node):
+        return self.ctx.float_ty(node.n)
+
+    def visit_Assign(self, node):
+        value = self.visit(node.value)
+
+        def _assign_target(target, value):
+            if type(target) == ast.Name:
+                id = target.id
+                self.register[id] = value
+            elif type(target) == ast.Tuple:
+                for element, val in zip(target.elts, value):
+                    _assign_target(element, val)
+            else:
+                to_store = value
+                if hasattr(value.type, 'pointee'):
+                    to_store = builder.load(value)
+
+                target = self.visit(target)
+                self.builder.store(to_store, target)
+
+        for target in node.targets:
+            _assign_target(target, value)
+
+    def visit_Tuple(self, node):
+        return (self.visit(element) for element in node.elts)
+
+    def visit_BinOp(self, node):
+        operator = self.visit(node.op)
+        return operator(self.visit(node.left), self.visit(node.right))
+
+    def visit_BoolOp(self, node):
+        operator = self.visit(node.op)
+        values = [self.visit(value) for value in node.values]
+        ret_val = values[0]
+        for value in values[1:]:
```

Comment:
can bool op every be anything other than unary or binary?

Reply:
no, but the abstract grammar (https://docs.python.org/3/library/ast.html) specifies values is a tuple. Not sure if there is a different interpretation

Reply:
that's ok. but you can do:
```
assert len(node.values) == 2
lhs, rhs = (self.visit(value) for value in node.values)
...
```
The current version is OK (more generic), I was just wondering if it's ever needed